File: lab6/c/hashmap.c
#define __USE_XOPEN_EXTENDED
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <assert.h>
#include <string.h>

#include <math.h>

#include "hashmap.h"
#include "general.h"

typedef char *hashmap_mkey_t;

// This is a cell struct assuming Open Addressing
// You will need alternative data-structurs for separate chaining
typedef struct
{ // hash-table entry
  hashmap_mkey_t key;
  hashmap_value_t value;
  enum {empty, in_use, deleted} state;
} cell;

struct _hashmap_t
{
  cell *cells;
  size_t size; // cell cells [table_size];
  size_t num_entries; // number of cells in_use
  hashing_mode_t mode;
  //TODO add anything else that you need
  unsigned long long collision;
};

typedef size_t hashcode_t; // unsigned integer type <= size_t !

// Can be redefined if Value_Type changes
int compare_keys(hashmap_key_t a, hashmap_key_t b){
  return strcmp(a,b);
}

// Helper functions for finding prime numbers 
bool isPrime (size_t n)
{
  for (size_t i = 2; i*i <= n; i++)
    if (n % i == 0)
      return false;
  return true;
}

size_t nextPrime(size_t n)
{
  for (; !isPrime(n); n++);
  return n;
}
// h = x0*a^(k-1) + x1*a^(k-2) + ... + x(k-2)*a + x(k-1)
// h = |ak + b| mod N
hashcode_t hashFunc(hashmap_key_t k, size_t N)
{
  int temp = 1; int a = 41;
  size_t len = strlen(k);
  long long h = 0;

  //printf("%s, len: %zu, k[%zu]: %lld\n", k, len, len - 1, h);
  size_t i=len;
  while (i>0) {
    --i;
    h += k[i] * temp;
    //printf("h += k[%zu](%d) * %d, h = %lld\n", i, k[i], a, h);
    temp *= a;
  }

  //printf("%lld\n", h);
  h = (hashcode_t)(labs(h) % N);
  return h;
}


bool isLinearProbing(hashing_mode_t mode){
  return mode == HASH_1_LINEAR_PROBING || mode == HASH_2_LINEAR_PROBING;
}
bool isQuadraticProbing(hashing_mode_t mode){
  return mode == HASH_1_QUADRATIC_PROBING || mode == HASH_2_QUADRATIC_PROBING;
}
bool isDoubleHashing(hashing_mode_t mode){
  return mode == HASH_1_DOUBLE_HASHING || mode == HASH_2_DOUBLE_HASHING;
}


#define DOUBLE_HASH_Q 7919

hashmap_t* hashmap_new(size_t size, hashing_mode_t mode)
{
  if (size<5) size=5;
  size=nextPrime(size);

  if (isDoubleHashing(mode) && size <= DOUBLE_HASH_Q ) {
    assert(isPrime(DOUBLE_HASH_Q));
    size = nextPrime(DOUBLE_HASH_Q + 1);
  }

  hashmap_t* set = CALLOC(hashmap_t,1);
  set->cells = CALLOC(cell,size);
  set->size = size;
  set->num_entries = 0;
  set->collision = 0;
  set->mode = mode;

  for(size_t i = 0; i < size; i++){
    set->cells[i].state = empty;
  }
  return set;

}

void hashmap_delete(hashmap_t* set)
{
  for(size_t i=0;i<set->size;i++){
    if(set->cells[i].state == in_use){
      free(set->cells[i].key);
    }
  }
  free(set->cells);
  free(set);
}

size_t hashmap_get_size(hashmap_t const* set){ return set->num_entries; }

// insertion

void insertCell(hashmap_key_t k, hashmap_value_t v, hashmap_t* set, size_t pos)
{
 set->cells[pos].key = strdup(k);
 set->cells[pos].value=v;
 set->cells[pos].state = in_use;
 set->num_entries ++;
  return;
}

hashmap_t* rehash(hashmap_t* set)
{
  size_t N =set->size;
  size_t newSize = nextPrime(2*N);
  hashmap_t* new_set = hashmap_new(newSize, set->mode);
  for(size_t i = 0; i < N; i++){
    if(set->cells[i].state == in_use)
      hashmap_insert(set->cells[i].key, set->cells[i].value, new_set);
  }
  hashmap_delete(set);
  return new_set;
}


hashmap_t* hashmap_insert (hashmap_key_t key, hashmap_value_t value, hashmap_t* set)
{
  if(hashmap_contains(key, set)) error("Duplicate key %s", key); // detect duplicate

  size_t N = set->size;
  if(   set->num_entries >= N
    || (isQuadraticProbing(set->mode) && set->num_entries >= N / 2 ))
  {
    set = rehash(set);
  }
  N = set->size;

  //printf("INSERT %s -> %zu\n", key,value);

  hashcode_t pos = hashFunc(key, N);

  // insert into empty cell
  if(set->cells[pos].state != in_use)
    insertCell(key, value, set, pos);

  else{
    if(isLinearProbing(set->mode)) // linear probing: A[(i+j)mod N]
    {

      for(size_t j = 1; j < N; j++){
        size_t look = (pos + j) % N;
       set->collision ++;
        if(set->cells[look].state != in_use){
          insertCell(key, value, set, look);
          break;
        }
      }
    }
    else if(isQuadraticProbing(set->mode)) // quadratic probing: A[(i+j^2)mod N]
    {

      for(size_t j = 1; j < N; j++){
        size_t look = (pos + j*j) % N;

       set->collision ++;
        if(set->cells[look].state != in_use){
          insertCell(key, value, set, look);
          break;
        }
      }
    }
    else if(isDoubleHashing(set->mode)) // double hash: A[(i+j*h'(k))mod N] h'(k)=q-(k mod q)
    {
      size_t q = DOUBLE_HASH_Q; // q<N a prime number
      assert(q<N);

      size_t temp = q - pos % q;

      for(size_t j = 1; j < N; j++){
        size_t look = (pos + j*temp) % N;

       set->collision ++;
        if(set->cells[look].state != in_use){
          insertCell(key, value, set, look);
          break;
        }
      }
    } else assert(false);
  }
  return set;
}

bool hashmap_lookup(hashmap_key_t key, hashmap_value_t *value, hashmap_t const* set) {
  size_t N =set->size;
  hashcode_t pos = hashFunc(key, N);

  if(set->cells[pos].state == in_use
        && (strcmp(set->cells[pos].key, key) == 0))
  {
    if (value) *value=set->cells[pos].value;
    return true;
  }
  else if(isLinearProbing(set->mode)) // linear probing
  {
    for(size_t j = 1; j < N; j++){
      size_t look = (pos + j) % N;
      if(set->cells[look].state == empty)
        return false;
      if(set->cells[look].state == in_use
        && (strcmp(set->cells[look].key, key) == 0))
      {
        if (value) *value=set->cells[look].value;
        return true;
      }
    }
  }
  else if(isQuadraticProbing(set->mode)) // quadratic probing
  {
    for(size_t j = 1; j < N; j++){
      size_t look = (pos + j*j) % N;

      if(set->cells[look].state == empty)
        return false;
      if(set->cells[look].state == in_use &&
        (strcmp(set->cells[look].key, key) == 0))
      {
        if (value) *value=set->cells[look].value;
        return true;
      }
    }
  }
  else if(isDoubleHashing(set->mode)) // double hash
  {
    size_t q = DOUBLE_HASH_Q;
    size_t temp = q - pos % q;
    assert(q<N);

    for(size_t j = 1; j < N; j++){
      size_t look = (pos + j*temp) % N;

      if(set->cells[look].state == empty)
        return false;
      if(set->cells[look].state == in_use &&
        (strcmp(set->cells[look].key, key) == 0))
      {
        if (value) *value=set->cells[look].value;
        return true;
      }
    }
  }
  return false;
}


bool hashmap_contains (hashmap_key_t key, hashmap_t const* set)
{
  return hashmap_lookup(key,NULL,set);
}

void hashmap_print_set (hashmap_t const* set)
{
  for(size_t i = 0; i < set->size; i++){
    if(set->cells[i].state == in_use)
      printf("Cell %zu: %s=>%zu\n", i, set->cells[i].key, set->cells[i].value);
    if(set->cells[i].state == empty)
      printf("Cell %zu: empty\n", i);
    if(set->cells[i].state == deleted)
      printf("Cell %zu: deleted\n", i);
  }
}

void hashmap_print_stats (hashmap_t const* set)
{
  printf("Collision times: %llu\n", set->collision);
  printf("Entry number: %zu\n", set->num_entries);
  printf("Average collisions per access: %lf\n", (double)set->collision / (double)set->num_entries);
}

File: lab6/python/hashmap.py
from enum import Enum
import general

DOUBLE_HASH_Q = 7919
    
class hashmap_t:
    def __init__(self, size, mode):
        if (size < 5):
            size = 5
        size = self.nextPrime(size)
        
        if (self.isDoubleHashing(mode) and size <= DOUBLE_HASH_Q):
            assert(self.isPrime(DOUBLE_HASH_Q))
            size = self.nextPrime(DOUBLE_HASH_Q + 1)
        
        self.size = size
        self.cells = self.initialise_cell_array()
        self.mode = mode
        
    def initialise_cell_array(self):
        self.num_entries = 0
        self.collision = 0
        cells = []

        for i in range(self.size):
            new_cell = cell()
            new_cell.state = state.empty
            cells.append(new_cell)
        return cells
        
                
    # Helper functions for finding prime numbers
    def isPrime(self, n):
        i = 2
        while (i * i < n):
            if (n % i == 0):
                return False
            i = i + 1
        return True
        
    def nextPrime(self, n):
        while (not self.isPrime(n)):
            n = n + 1
        return n
        
    def hashFunc(self, k, N):
        temp = 1
        a = 41
        string_len = len(k)
        h = 0
        i = string_len - 1
        while i >= 0:
            h = h + ord(k[i]) * temp
            temp = temp * a
            i = i - 1
            
        h = abs(h) % N
        return h
        
    def isLinearProbing(self, mode):
        return (mode == HashingModes.HASH_1_LINEAR_PROBING or mode == HashingModes.HASH_2_LINEAR_PROBING)
        
    def isQuadraticProbing(self, mode):
        return (mode == HashingModes.HASH_1_QUADRATIC_PROBING or mode == HashingModes.HASH_2_QUADRATIC_PROBING)
        
    def isDoubleHashing(self, mode):
        return (mode == HashingModes.HASH_1_DOUBLE_HASHING or mode == HashingModes.HASH_2_DOUBLE_HASHING)
    
    def hashmap_get_size(self):
        return self.num_entries
        
    # insertion
    
    def insertCell(self, k, v, pos):
        self.cells[pos].key = k
        self.cells[pos].value = v
        self.cells[pos].state = state.in_use
        self.num_entries = self.num_entries + 1
        
    def rehash(self):
        N = self.size
        newSize = self.nextPrime(2 * N)
        self.size = newSize
        old_cells = self.cells.copy()
        
        self.cells = self.initialise_cell_array()
        for i in range(N):
            if (old_cells[i].state == state.in_use):
                self.hashmap_insert(old_cells[i].key,old_cells[i].value)
                        

    def hashmap_insert(self, key, value):
        if self.hashmap_contains(key):
            general.error("Duplicate key %s", key) # detect duplicate
            
        N = self.size
        if (self.num_entries >= N or (self.isQuadraticProbing(self.mode) and self.num_entries >= N/2)):
            self.rehash()
        N = self.size
            
        pos = self.hashFunc(key, N)
        
        # insert into empty cell
        if (self.cells[pos].state != state.in_use):
            self.insertCell(key, value, pos)
        else:
            if self.isLinearProbing(self.mode): # linear probing: A[(i + j)mod N]
                for j in range(1, N):
                    look = (pos + j) % N
                    self.collision = self.collision + 1
                    if (self.cells[look].state != state.in_use):
                        self.insertCell(key, value, look)
                        break
            elif self.isQuadraticProbing(self.mode): # quadratic probing : A[(i + j^2)mod N]
                for j in range (1, N):
                    look = (pos + j*j) % N
                    self.collision = self.collision + 1
                    if (self.cells[look].state != state.in_use):
                        self.insertCell(key, value, look)
                        break
            elif self.isDoubleHashing(self.mode): # double hash: A[(i+j*h'(k))mod N] h'(k)=q-(k mod q)
                q = DOUBLE_HASH_Q # q<N a prime number
                assert(q<N)
                temp = q - pos % q
                
                for j in range(1, N):
                    look = (pos + j * temp) % N
                    self.collision = self.collision + 1
                    if (self.cells[look].state != state.in_use):
                        self.insertCell(key, value, look)
                        break
        
    def hashmap_lookup(self, key):
        N = self.size
        pos = self.hashFunc(key, N)
        
        value = None
        
        if (self.cells[pos].state == state.in_use and self.cells[pos].key == key):
            if (value == None):
                value = self.cells[pos].value
                return value;
            
        elif self.isLinearProbing(self.mode):
            for j in range(1, N):
                look = (pos + j) % N
                if (self.cells[look].state == state.empty):
                    return None
                if (self.cells[look].state == state.in_use and self.cells[look].key == key):
                    if (value == None):
                        value = self.cells[look].value
                        return value
        elif self.isQuadraticProbing(self.mode):
            for j in range(1, N):
                look = (pos + j*j) % N
                if (self.cells[look].state == state.empty):
                    return None
                if (self.cells[look].state == state.in_use and self.cells[look].key == key):
                    if (value == None):
                        value = self.cells[look].value
                        return value
        elif self.isDoubleHashing(self.mode):
            q = DOUBLE_HASH_Q
            temp = q - pos % q
            assert(q<N)
            for j in range(1, N):
                look = (pos + j*temp) % N
                if (self.cells[look].state == state.empty):
                    return None
                if (self.cells[look].state == state.in_use and self.cells[look].key == key):
                    if (value == None):
                        value = self.cells[look].value
                        return value
        return value
    
    def hashmap_contains(self, key):
        if (not self.hashmap_lookup(key) == None):
            return True
        return False
        
    def hashmap_print_set(self):
        for i in range(self.size):
            if self.cells[i].state == state.in_use:
                print("Cell %5d: %s => %d" % (i,  self.cells[i].key, self.cells[i].value))
            if (self.cells[i].state == state.empty):
                print("Cell %5d: empty" % i)
            if (self.cells[i].state == state.deleted):
                print("Cell %5d: deleted" % i)
        
    def hashmap_print_stats(self):
        print("Collision times: %d\n" % self.collision)
        print("Entry number: %d\n" % self.num_entries)
        print("Average collisions per access: %lf\n" % (self.collision/self.num_entries))
        
# This is a cell structure assuming Open Addressing
# It should contain and element that is the key and a state which is empty, in_use or deleted
# You will need alternative data-structures for separate chaining
class cell:
    def __init__(self):
        pass
        
class state(Enum):
    empty = 0
    in_use = 1
    deleted = 2
        
# Hashing Modes
class HashingModes(Enum):
    HASH_1_LINEAR_PROBING=0
    HASH_1_QUADRATIC_PROBING=1
    HASH_1_DOUBLE_HASHING=2
    HASH_1_SEPARATE_CHAINING=3
    HASH_2_LINEAR_PROBING=4
    HASH_2_QUADRATIC_PROBING=5
    HASH_2_DOUBLE_HASHING=6
    HASH_2_SEPARATE_CHAINING=7
